src/main/scala/sangria/marshalling/ArrayMapBuilder.scala

```diff
@@ -62,4 +63,40 @@ class ArrayMapBuilder[T](keys: Seq[String]) {
 
     builder.result()
   }
+
+  lazy val toIterator: Iterator[(String, T)] = {
+    new Iterator[(String, T)] {
+      var index = -1
+      var nextIndex = 0
+
+      @tailrec def nextIndex(current: Int): Option[Int] = {
+        val next = current + 1
+        if (next >= elements.length) None
+        else if (indexesSet.contains(next)) Some(next)
+        else nextIndex(next)
+      }
+
+      override def hasNext: Boolean = {
+        nextIndex(index) match {
```

Comment:
Do you think it would be possible to optimise `hasNext` in case it is called multiple times (without a call to `next`). In this scenario, `def nextIndex` has a potential to cause a lot of repeated comparisons for big sparse field lists (which might be quite common). Maybe we can cache it's result somehow?

Reply:
My assumption is that a user is always calling `next()` after `hasNext()` (otherwise we cannot make any progress).
But I'm quite a newbie in the iterator business. I'd be glad to lean more about it if you have more info 😉